core-bundle/src/Controller/FrontendModule/FeedReaderController.php

```diff
@@ -0,0 +1,114 @@
+
+use Contao\Config;
+use Contao\CoreBundle\Exception\PageNotFoundException;
+use Contao\CoreBundle\ServiceAnnotation\FrontendModule;
+use Contao\CoreBundle\Twig\FragmentTemplate;
+use Contao\Environment;
+use Contao\ModuleModel;
+use Contao\Pagination;
+use Contao\StringUtil;
+use FeedIo\Feed;
+use FeedIo\Feed\Item;
+use FeedIo\FeedInterface;
+use FeedIo\FeedIo;
+use Psr\Log\LoggerInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Contracts\Cache\CacheInterface;
+use Symfony\Contracts\Cache\ItemInterface;
+
+/**
+ * @FrontendModule(category="miscellaneous")
+ */
+class FeedReaderController extends AbstractFrontendModuleController
+{
+    public function __construct(private readonly FeedIo $feedIo, private readonly LoggerInterface $logger, private readonly CacheInterface $cache)
+    {
+    }
+
+    protected function getResponse(FragmentTemplate $template, ModuleModel $model, Request $request): Response
+    {
+        $this->initializeContaoFramework();
+
+        $feeds = [];
+
+        foreach (StringUtil::trimsplit('[\n\t ]', trim($model->rss_feed)) as $url) {
+            try {
+                $feed = $this->cache->get('feed_reader_'.$model->id.'_'.md5($url), function (ItemInterface $item) use ($url, $model) {
+                    $readerResult = $this->feedIo->read($url, new Feed());
+
+                    if ($model->rss_cache > 0) {
+                        $item->expiresAfter($model->rss_cache);
+                    }
+
+                    return $readerResult->getFeed();
+                });
+            } catch (\Exception $exception) {
+                $feed = null;
+                $this->logger->error(sprintf('Could not read feed %s: %s', $url, $exception->getMessage()));
+                continue;
+            }
+
+            if ($feed instanceof FeedInterface) {
+                $feeds[] = $feed;
+            }
+        }
+
+        $template->set('feeds', $feeds);
+
+        // Take the configured amount of items from each feed and merge them into one list
+        /** @var list<array{feed: Feed, item: Item}> $elements */
+        $elements = array_merge(
+            ...array_map(
+                static fn (FeedInterface $feed): array => array_map(
+                    static fn (Item $item) => [
+                        'feed' => $feed,
+                        'item' => $item,
+                    ],
+                    \array_slice([...$feed], $model->skipFirst, $model->numberOfItems ?: null)
+                ),
+                $feeds
+            )
+        );
+
+        usort($elements, static fn (array $a, array $b): int => $a['item']->getLastModified() <=> $b['item']->getLastModified());
+
+        if ($model->perPage > 0) {
+            $param = 'page_r'.$model->id;
+            $page = $request->query->getInt($param, 1);
+            $config = $this->container->get('contao.framework')->getAdapter(Config::class);
```

Comment:
This is wrong. The framework should be injected instead.

Reply:
Even though we explicitly require that service in the parent class? https://github.com/contao/contao/blob/0f81487b38bc93af15abc341ea05ea5400a297f7/core-bundle/src/Controller/AbstractController.php#L33

Reply:
We also require the `Logger` class in the abstract controller and still it is injected in line 36 instead of fetched via `$this->container->get('logger')`.

https://github.com/contao/contao/blob/0f81487b38bc93af15abc341ea05ea5400a297f7/core-bundle/src/Controller/AbstractController.php#L35

Not sure what is best practice here. 🤔

Reply:
It seems we have a shortcut method to get Contao adapters in controllers, so I guess 2965c571d3930213f6b9d514ba6f0c01e9c32a03 is the correct way to go.

Reply:
> Even though we explicitly require that service in the parent class

In general: yes, that's the best practice. This way you can alter things in the base class (e.g. refactor stuff and thereby remove an injected service) without breaking stuff.